pandas/util/_decorators.py

```diff
@@ -247,6 +247,36 @@ def wrapper(*args, **kwargs) -> Callable[..., Any]:
     return decorate
 
 
+def doc(*args: Union[str, Callable], **kwargs: str) -> Callable:
+    """
+    A decorator take docstring templates, concatenate them and perform string
+    substitution on it.
+
+    This decorator should be robust even if func.__doc__ is None.
+    """
+
+    def decorator(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(*args, **kwargs) -> Callable:
+            return func(*args, **kwargs)
+
+        templates = [func.__doc__ if func.__doc__ else ""]
+        for arg in args:
+            if isinstance(arg, str):
+                templates.append(arg)
+            elif hasattr(arg, "_docstr_template"):
+                templates.append(arg._docstr_template)  # type: ignore
+            elif arg.__doc__:
+                templates.append(arg.__doc__)
+
+        wrapper._docstr_template = "".join(dedent(t) for t in templates)  # type: ignore
```

Comment:
The `doc` decorator will format docstring template and use the formatted string to replace `__doc__`. Using the decorator will cause us to lose the original template. 

Take this fake code as an example:
```python
@doc(method='cummax', operation='maximum')
def cummax(whatever):
    """
    This is the {method} method.

    It computes the cumulative {operation}.
    """

@doc(cummax, method='cummin', operation='minimum')
def cummin(whatever):
    pass
```

In `cummin`, we would like to re-use the docstring template from `cummax`. However, when calling `@doc()` for `cummax`, the `__doc__` will no longer be the template. So, we might want to save the original template somewhere and use it later.

I think it is a good idea to commenting this code because, at least for me, this is not intuitive. 
My questions are: 
1. Will it be better to put the explanation in docstring or in the code block?
2. Is there an example (about formatting) that I could follow?

Reply:
> Can you explain / add a comment why the `_docstr_template` is needed?

@jorisvandenbossche I have [add more detail doc](https://github.com/pandas-dev/pandas/pull/31060/commits/88c6bf2830647ea65cbcfd13943e65afb5f86dee) under the decorator. May I have more feedback please?